### PR TITLE
api: mark internal listener as implemented and change internal listener binding

### PR DESCRIPTION
**Pull request by kyessenov**
Signed-off-by: Kuat Yessenov <kuat@google.com>

Commit Message: Expose internal listener functionality. Marks the fields as implemented. Adds the following functional changes prior to publishing:
* do not use the `address` field per https://github.com/envoyproxy/envoy/pull/21132, and instead use the listener `name` field as originally expected.
* add an `endpoint_id` field to upstream address to fix https://github.com/envoyproxy/envoy/pull/22420; it's not possible to have multiple internal addresses to the same listener in a single pool without some extra identifier beyond the listener name, because pools identify addresses using `asString()`.
* fixes https://github.com/envoyproxy/envoy/issues/20665 by using stat prefix `listener.envoy_internal_<listener_name>`.

Adds a few sample configs (and fixes https://github.com/envoyproxy/envoy/issues/20335):
* minimal two chained `tcp_proxy` listeners;
* encap HTTP in HTTP CONNECT;
* decap HTTP in HTTP CONNECT;

Risk Level: low, hidden prior
Testing: yes
Docs Changes: yes
Release Notes: yes

### CODE REVIEW

**Comment by phlax** [comment]
@adisuissa would you mind taking a look

/wait-any

File: api/envoy/config/core/v3/address.proto
Comment (adisuissa) [inline api/envoy/config/core/v3/address.proto:41]: Please keep the comment describing the field.

File: api/envoy/config/listener/v3/listener.proto
Comment (adisuissa) [inline api/envoy/config/listener/v3/listener.proto:335]:
```suggestion
  // [#comment: TODO(lambdai): https://github.com/envoyproxy/envoy/issues/15372]
```

**Review by adisuissa — CHANGES_REQUESTED** [review]
Thanks for cleaning this up, left a few minor comments.
/wait

File: api/envoy/config/core/v3/address.proto
Comment (kyessenov) [reply api/envoy/config/core/v3/address.proto:41]: This comment is not true, address has nothing to do with the listener name.

File: api/envoy/config/listener/v3/listener.proto
Comment (kyessenov) [reply api/envoy/config/listener/v3/listener.proto]: Done.

File: api/envoy/config/core/v3/address.proto
Comment (adisuissa) [reply api/envoy/config/core/v3/address.proto:41]: IIUC the value here needs to point to another listener.
Would it be possible to add a reference to the field that this value will be matched to?

Comment (soulxu) [reply api/envoy/config/core/v3/address.proto:41]: seems like it changed what the value expect to from the beginning. Maybe worth to ask again whether this field still name correctly, this can be a chance to change the name before expose to the end user.

**Comment by alyssawilk** [comment]
Looks like a simple API change: throwing over to @htuch as Adi is out.
@kyessenov should this also have a release note as well?

**Review by adisuissa — DISMISSED** [review]
/lgtm api,
thanks!

**Comment by adisuissa** [comment]
@kyessenov following Alyssa's comment, if the `server_listener_name` field now implemented, then a release note should be added.

**Comment by alyssawilk** [comment]
/wait on review notes - I'm happy to do that pass.  I think if we don't have example configs we're going to want that as part of making this non-hidden.

File: api/envoy/config/core/v3/address.proto
Comment (alyssawilk) [inline api/envoy/config/core/v3/address.proto:146]: we should have comments here and above explaining what the field is for?

**Comment by kyessenov** [comment]
Affected by https://github.com/envoyproxy/envoy/issues/22483.

**Comment by kyessenov** [comment]
Per https://github.com/envoyproxy/envoy/pull/21132, I'm cleaning up the usage of `address` to replace with a field in `InternalListener` instead.

I'm also making a provision to address https://github.com/envoyproxy/envoy/pull/22420 with a dedicated field in upstream address for a port-like identifier.

File: api/envoy/config/core/v3/address.proto
Comment (kyessenov) [reply api/envoy/config/core/v3/address.proto:41]: Yes, I agree we need to fix this. I think we should use a dedicated field in InternalListener instead of internal address in the generic `address` field in Listener. WDYT?

**Comment by kyessenov** [comment]
I'll add a release note and update config samples. There're several issues that could be resolved in this PR, so I'm trying to be systematic before exposing this.

File: api/envoy/config/core/v3/address.proto
Comment (kyessenov) [reply api/envoy/config/core/v3/address.proto:146]: Done.

File: api/envoy/config/core/v3/address.proto
Comment (kyessenov) [reply api/envoy/config/core/v3/address.proto:41]: The simplest thing is to just really use the listener name. That keeps LDS to a minimum and I can't think of any reason why more flexibility would be needed here (multiple addresses don't work anyways, but can always be added to `InternalListener`).

Comment (soulxu) [reply api/envoy/config/core/v3/address.proto:41]: I'm not sure the usecase of using multiple internal addresses, it seems useless. I guess the multiple addresses usecase is internal address mix with IP address. From this point of view, just use the listener name should be fine.

File: api/envoy/config/core/v3/address.proto
Comment (soulxu) [inline api/envoy/config/core/v3/address.proto:42]: so this is only used for upstream endpoint?

File: api/envoy/config/core/v3/address.proto
Comment (soulxu) [inline api/envoy/config/core/v3/address.proto:47]: Is this kind of multiple internal addresses? if yes, then another option is just to support multiple internal addresses here, then you needn't add `endpoint_id` field for multiple endpoints, just specify multiple internal addresses. but yes, you can't limit the internal address to the listener name anymore.

File: configs/encapsulate_http_in_http2_connect.yaml
Comment (phlax) [inline configs/encapsulate_http_in_http2_connect.yaml:3]: is admin required for this config? if not can we remove - i have tried to make the configs contain the least amount of configuration possible, esp when it comes to including admin

im also wondering the purpose of including these configs here - and whether this would be better served by adding a tunneling sandbox

File: api/envoy/config/core/v3/address.proto
Comment (soulxu) [reply api/envoy/config/core/v3/address.proto:41]: > I'm not sure the usecase of using multiple internal addresses, it seems useless. I guess the multiple addresses usecase is internal address mix with IP address. From this point of view, just use the listener name should be fine.

suddenly found this may be the multiple internal addresses usecase https://github.com/envoyproxy/envoy/pull/22274/files#r935253240

File: source/server/connection_handler_impl.cc
Comment (soulxu) [inline source/server/connection_handler_impl.cc:134]: the line 132 should be already ensure this is internal address

File: source/server/listener_impl.cc
Comment (soulxu) [inline source/server/listener_impl.cc:552]: The name isn't a good choice also, since 
https://github.com/envoyproxy/envoy/blob/a53743b20351288a37a61811f27bcda463dd32ad/source/server/listener_manager_impl.cc#L381-L383

And when we support multiple addresses which mix internal and IP address, log the addresses is better.

File: api/envoy/config/core/v3/address.proto
Comment (soulxu) [reply api/envoy/config/core/v3/address.proto:47]: after looking at another PR, multiple addresses will bind the number of listener addresses and the number of endpoints together, I guess that isn't what people want.

File: source/server/listener_impl.cc
Comment (soulxu) [inline source/server/listener_impl.cc:535]: this check https://github.com/envoyproxy/envoy/blob/c4837b52cebc46022e536061755ee29797c97c32/source/server/listener_manager_impl.cc#L386-L394

is useless anymore.

or we can move this check to 
https://github.com/envoyproxy/envoy/blob/c4837b52cebc46022e536061755ee29797c97c32/source/server/listener_manager_impl.cc#L387
and replace the original check, then it will avoid ListenerImpl's construct to do the useless parsing for the internal addresses
https://github.com/envoyproxy/envoy/pull/22274/files#diff-b0d5e1a1317df8e22a556ee74d6f728379c5f85ce640a9404acc14bfe8192ad3R371

File: source/server/listener_impl.cc
Comment (soulxu) [inline source/server/listener_impl.cc:371]: there should be a case, `config.has_internal_listener()` is false, but internal address is specified in the address field.

File: configs/encapsulate_http_in_http2_connect.yaml
Comment (phlax) [reply configs/encapsulate_http_in_http2_connect.yaml:3]: thinking on this further - why not just put these in the docs and add as literalinclude examples - it would be simpler than adding a sandbox, but at least they would be visible/used

**Review by alyssawilk — COMMENTED** [review]
This is shaping up really well - thanks for your patience on iterating :-)
/wait

File: api/envoy/config/core/v3/address.proto
Comment (kyessenov) [reply api/envoy/config/core/v3/address.proto:47]: Correct, we want to maximize reuse of the internal listeners - e.g. tunneling shouldn't be coupled to the origin of the traffic. So that implies a possibility of multiple endpoints to the same internal listener in one pool. Upstream de-duplicates endpoints by computing `asString` so it breaks there.

File: api/envoy/config/core/v3/address.proto
Comment (kyessenov) [reply api/envoy/config/core/v3/address.proto:46]: It is an opaque identifier. It's only used so that `asString` is different when `server_listener_name` is the same. Otherwise, EDS will delete repeated endpoints to the same internal listener.

File: configs/encapsulate_http_in_http2_connect.yaml
Comment (kyessenov) [reply configs/encapsulate_http_in_http2_connect.yaml:3]: 1. I'll remove admin, I thought it's required.
2. They are too large as literal include because they need the entirety to be meaningful. Snippets work best for isolated config samples.
3. Yes, users will copy paste so having a working example is really helpful.

File: source/server/listener_impl.cc
Comment (kyessenov) [reply source/server/listener_impl.cc:552]: Address is empty for internal listener I think in this branch.

File: configs/encapsulate_http_in_http2_connect.yaml
Comment (kyessenov) [reply configs/encapsulate_http_in_http2_connect.yaml:3]: Let's just do both, add snippets and fully worked out examples.

File: source/server/listener_impl.cc
Comment (kyessenov) [reply source/server/listener_impl.cc:371]: Oh you mean we don't check for that early enough? We should throw in buildInternalListener but that might be late.

Comment (soulxu) [reply source/server/listener_impl.cc:371]: yea, I saw you add the check here https://github.com/envoyproxy/envoy/pull/22274/files#diff-b0d5e1a1317df8e22a556ee74d6f728379c5f85ce640a9404acc14bfe8192ad3R569, that is what I said.

File: source/server/listener_impl.cc
Comment (soulxu) [reply source/server/listener_impl.cc]: You append an internal addresses here https://github.com/envoyproxy/envoy/pull/22274/files#diff-b0d5e1a1317df8e22a556ee74d6f728379c5f85ce640a9404acc14bfe8192ad3R366-R367, so it shouldn't be empty

File: api/envoy/config/core/v3/address.proto
Comment (soulxu) [inline api/envoy/config/core/v3/address.proto:36]: Maybe worth to note that this is only used for upstream endpoint address.

File: api/envoy/config/listener/v3/listener.proto
Comment (soulxu) [inline api/envoy/config/listener/v3/listener.proto:106]: Also maybe worth to note, this will make the listener use the listener name as the internal address. And when this is internal listener, the address field should keep empty.

File: api/envoy/config/listener/v3/listener.proto
Comment (soulxu) [inline api/envoy/config/listener/v3/listener.proto:347]: Also maybe worth to note, this will make the listener use the listener name as the internal address. And when this is internal listener, the address field should keep empty.

File: source/server/listener_impl.cc
Comment (soulxu) [inline source/server/listener_impl.cc:538]: looks like missing a testcase for this case.

**Review by soulxu — COMMENTED** [review]
overall LGTM, only left few comments about doc and test

**Review by alyssawilk — COMMENTED** [review]
LGTM modulo outstanding comments
/wait

File: configs/encapsulate_http_in_http2_connect.yaml
Comment (kyessenov) [reply configs/encapsulate_http_in_http2_connect.yaml]: Resolving since we can have both snippets and full configs.

File: api/envoy/config/core/v3/address.proto
Comment (kyessenov) [reply api/envoy/config/core/v3/address.proto:36]: Added.

File: api/envoy/config/listener/v3/listener.proto
Comment (kyessenov) [reply api/envoy/config/listener/v3/listener.proto:347]: Added.

File: source/server/listener_impl.cc
Comment (kyessenov) [reply source/server/listener_impl.cc]: Internal listeners are always deliberately named since that's the only way to refer to them in EnvoyInternalAddress. I think we can rework this if/when we add multiple address support for internal listeners.

File: source/server/listener_impl.cc
Comment (kyessenov) [reply source/server/listener_impl.cc:538]: Good catch. This is too late to check for `api_listener` so I moved it earlier.

File: api/envoy/config/core/v3/address.proto
Comment (adisuissa) [inline api/envoy/config/core/v3/address.proto:46]: Is empty string the default value here, or some unique name for the endpoint?
Is there a requirement that each endpoint id will be unique?

**Review by adisuissa — COMMENTED** [review]
Thanks for working on this, exciting feature!

File: api/envoy/config/core/v3/address.proto
Comment (kyessenov) [reply api/envoy/config/core/v3/address.proto:35]: I think that comment is stale, this address is already available in both places.

File: docs/root/configuration/other_features/internal_listener.rst
Comment (kyessenov) [reply docs/root/configuration/other_features/internal_listener.rst:24]: It'll be an upstream connect error, same as if server is not listening.

File: api/envoy/config/core/v3/address.proto
Comment (kyessenov) [reply api/envoy/config/core/v3/address.proto:46]: Yes, the empty string is the default. There's no requirement per-se, but it's de-duplicated in the same way as when EDS has duplicate IP addresses.

**Comment by soulxu** [comment]
LGTM

File: api/envoy/config/core/v3/address.proto
Comment (lambdai) [reply api/envoy/config/core/v3/address.proto:47]: Is there any typical value of endpoint_id? e.g. the actually L4 address?

File: source/common/network/address_impl.cc
Comment (lambdai) [inline source/common/network/address_impl.cc:411]: Given the endpoint_id in the real world contains ":" (ipv6 address and port). It might be good to use another separator here.

**Review by lambdai — COMMENTED** [review]
Thank you!

Only nits

File: api/envoy/config/core/v3/address.proto
Comment (stevenctl) [reply api/envoy/config/core/v3/address.proto:47]: @lambdai yes exactly

File: api/envoy/config/core/v3/address.proto
Comment (adisuissa) [reply api/envoy/config/core/v3/address.proto:35]: IIUC, the 2nd TODO (`TODO(asraa)`) seems to still be relevant (the code is still there:
https://github.com/envoyproxy/envoy/blob/5e4c50b0c50db8877cf7197c1b6c1853e27e783f/test/server/server_fuzz_test.cc#L29-L31
).

If it is stale, then the test code should be fixed. If it is not, then the `[#comment` tag should be around that TODO.

File: api/envoy/config/core/v3/address.proto
Comment (adisuissa) [reply api/envoy/config/core/v3/address.proto:46]: Ah, I see. I think it may be beneficial to state that this is used for tracking purposes (opposed to an ID that will be used as a foreign key).

File: api/envoy/config/core/v3/address.proto
Comment (kyessenov) [reply api/envoy/config/core/v3/address.proto:47]: I think it'd be confusing to say "L4 address should be here". We don't say a listener corresponds to VIP, or a cluster corresponds to a VIP. That's too prescriptive. I will add an example of the usage.

File: api/envoy/config/core/v3/address.proto
Comment (kyessenov) [reply api/envoy/config/core/v3/address.proto:35]: Ah, missed that, thanks.

File: api/envoy/config/core/v3/address.proto
Comment (kyessenov) [reply api/envoy/config/core/v3/address.proto:46]: Done.

File: source/common/network/address_impl.cc
Comment (kyessenov) [reply source/common/network/address_impl.cc:411]: Sure, will use slash instead.

**Comment by kyessenov** [comment]
/retest
(emsdk download returned 500)

**Review by adisuissa — COMMENTED** [review]
/lgtm api

File: docs/root/configuration/other_features/internal_listener.rst
Comment (snowp) [inline docs/root/configuration/other_features/internal_listener.rst:64]: What happens if you don't use the endpoint id in case? What do you end up losing out on? Or does it just not work?

**Review by snowp — CHANGES_REQUESTED** [review]
Not familiar enough with the internal listener work to give a true in depth review, but high level this looks good to me, just a docs comment

File: docs/root/configuration/other_features/internal_listener.rst
Comment (kyessenov) [reply docs/root/configuration/other_features/internal_listener.rst:64]: It'll de-duplicate the internal addresses. The same happens for IP:PORTs, but not well-documented. This is due to `HostMap` using strings as a key for host tracking. Should we add this comment here?

Comment (snowp) [reply docs/root/configuration/other_features/internal_listener.rst:64]: I think it would be helpful to either make the language prescriptive (you should do this when multiple IPs to ensure X) or help the reader understand what this does if we want to talk about this as an optional thing to do

Comment (kyessenov) [reply docs/root/configuration/other_features/internal_listener.rst:64]: Done.

**Review by snowp — APPROVED** [review]
Thanks!

**Review by adisuissa — COMMENTED** [review]
/lgtm api

**Review by snowp — APPROVED** [review]

**Comment by kyessenov** [comment]
Thanks @lambdai for doing the actual work! Hopefully we can get more usage and value now that it's public and easily available.

**Pull request merged**
